Polyadic.Json/NewType/JsonNewTypeJsonConverterFactory.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Polyadic.Json.NewType
+{
+    internal sealed class JsonNewTypeJsonConverterFactory : JsonConverterFactory
+    {
+        public override bool CanConvert(Type typeToConvert) => Attribute.IsDefined(typeToConvert, typeof(JsonNewTypeAttribute));
+
+        public override JsonConverter CreateConverter(Type typeToConvert, JsonSerializerOptions options)
+        {
+            var metadata = GetMetadata(typeToConvert, options);
+            return (JsonConverter)Activator.CreateInstance(
+                typeof(JsonNewTypeConverter<,>).MakeGenericType(typeToConvert, metadata.InnerType),
+                metadata)!;
+        }
+
+        private static NewTypeMetadata GetMetadata(Type typeToConvert, JsonSerializerOptions options)
+        {
+            var constructor = FindConstructor(typeToConvert);
+            var innerType = constructor.GetParameters()[0].ParameterType;
+            var converter = options.GetConverter(innerType);
+            var valueProperty = FindValueProperty(typeToConvert, innerType);
+            return new NewTypeMetadata(innerType, converter, constructor, valueProperty);
+        }
+
+        private static ConstructorInfo FindConstructor(Type typeToConvert)
+        {
+            var constructorCandidates = typeToConvert.GetConstructors().Where(IsUnary).ToImmutableArray();
+            var markedConstructors = constructorCandidates.Where(IsMarkedAsJsonConstructor).ToImmutableArray();
+            return (markedConstructors.Length, constructorCandidates.Length) switch
+            {
+                (0, 0) => throw new JsonException($"No suitable constructors found for type '{typeToConvert}'. There must be at least one constructor with one parameter"),
+                (>1, _) => throw new JsonException($"Multiple constructors of type '{typeToConvert}' are marked with [JsonConstructor]"),
+                (0, >1) => throw new JsonException($"Multiple suitable constructors found for type '{typeToConvert}'. Choose one with [JsonConstructor]"),
+                (1, _) => markedConstructors[0],
```

Comment:
Is there a reason you use the array operator instead of First() ?

Reply:
44edcd7e993f8c034a31e140429837c0b1d65b30